lib/websocket/connection.js

```diff
@@ -0,0 +1,504 @@
+
+      if (secExtension !== null && secExtension !== permessageDeflate) {
+        failWebsocketConnection(controller, response.socket)
+        return
+      }
+
+      // 6. If the response includes a |Sec-WebSocket-Protocol| header field
+      //    and this header field indicates the use of a subprotocol that was
+      //    not present in the client's handshake (the server has indicated a
+      //    subprotocol not requested by the client), the client MUST _Fail
+      //    the WebSocket Connection_.
+      const secProtocol = response.headersList.get('Sec-WebSocket-Protocol')
+
+      if (secProtocol !== null && secProtocol !== request.headersList.get('Sec-WebSocket-Protocol')) {
+        failWebsocketConnection(controller, response.socket)
+        return
+      }
+
+      // processResponse is called when the "response’s header list has been received and initialized."
+      // once this happens, the connection is open
+      ws[kResponse] = response
+
+      whenConnectionEstablished(ws)
+
+      receiveData(ws)
+
+      socketClosed(ws)
+    }
+  })
+
+  return controller
+}
+
+/**
+ * @param {import('../fetch/index').Fetch} controller
+ * @param {import('net').Socket|import('tls').TLSSocket|undefined} socket
+ */
+function failWebsocketConnection (controller, socket) {
+  controller.abort()
+
+  if (socket && !socket.destroyed) {
+    socket.destroy()
+  }
+}
+
+/**
+ * @see https://websockets.spec.whatwg.org/#feedback-from-the-protocol
+ * @param {import('./websocket').WebSocket} ws
+ */
+function whenConnectionEstablished (ws) {
+  const { [kResponse]: response } = ws
+
+  // 1. Change the ready state to OPEN (1).
+  ws[kReadyState] = states.OPEN
+
+  // 2. Change the extensions attribute’s value to the extensions in use, if
+  //    it is not the null value.
+  // https://datatracker.ietf.org/doc/html/rfc6455#section-9.1
+  const extensions = response.headersList.get('sec-websocket-extensions')
+
+  if (extensions !== null) {
+    ws[kExtensions] = extensions
+  }
+
+  // 3. Change the protocol attribute’s value to the subprotocol in use, if
+  //    it is not the null value.
+  // https://datatracker.ietf.org/doc/html/rfc6455#section-1.9
+  const protocol = response.headersList.get('sec-websocket-protocol')
+
+  if (protocol !== null) {
+    ws[kProtocol] = protocol
+  }
+
+  // 4. Fire an event named open at the WebSocket object.
+  fireEvent('open', ws)
+
+  if (channels.open.hasSubscribers) {
+    channels.open.publish({
+      address: response.socket.address(),
+      protocol,
+      extensions
+    })
+  }
+}
+
+/**
+ * @see https://websockets.spec.whatwg.org/#feedback-from-the-protocol
+ * @param {import('./websocket').WebSocket} ws
+ */
+function receiveData (ws) {
+  const { [kResponse]: response } = ws
+
+  /** @type {WebsocketFrame|undefined} */
+  let frame
+
+  response.socket.on('data', (chunk) => {
+    const receivedFrame = WebsocketFrame.from(chunk)
```

Comment:
I'm pretty sure this still crashes if `chunk` is `Buffer.from([0x82, 0x7F])` as you assume that chunks can't be smaller than 8 bytes. That's a wrong assumption. Keep in mind that you can also receive a single byte so you have to wait until you have at least 2 bytes before reading the payload length, otherwise `buffer[1] & 0x7F` gives you 0 when it could actually be 127.